synapse/lib/layer.py

```diff
@@ -1345,7 +1381,155 @@ def getStorIndx(self, stortype, valu):
+                    continue
+
+                tag, prop = lkey[33:].decode().split(':')
+                valu, stortype = s_msgpack.un(lval)
+
+                buid = lkey[:32]
+
+                edit = (EDIT_TAGPROP_SET, (tag, prop, valu, None, stortype))
+                nodeedits[2].append(edit)
+                continue
+
+            logger.warning(f'unrecognized storage row: {flag}')
+
+        if nodeedits[0] is not None:
+            async for prop, valu in self.iterNodeData(nodeedits[0]):
+                edit = (EDIT_NODEDATA_SET, (prop, valu))
+                nodeedits[2].append(edit)
+
+            yield nodeedits
+
+    async def initUpstreamSync(self, url):
+        self.schedCoro(self._initUpstreamSync(url))
+
+    async def _initUpstreamSync(self, url):
+
+        while not self.isfini:
+
+            try:
+
+                async with await s_telepath.openurl(url) as proxy:
+
+                    offs = self.offsets.get(url)
+                    logger.warning(f'upstream sync connected ({url} offset={offs})')
+
+                    if offs == 0:
+                        offs = await proxy.getSpliceOffset()
+
+                        async for item in proxy.iterLayerSplices():
+                            await self.storNodeEdits([item], {})
+
+                        self.offsets.set(url, offs)
+
+                        waits = [v for k, v in self.upstreamwaits[url].items() if k <= offs]
+                        for wait in waits:
+                            [e.set() for e in wait]
+
+                    while not proxy.isfini:
+
+                        offs = self.offsets.get(url)
+
+                        # pump them into a queue so we can consume them in chunks
+                        q = asyncio.Queue(maxsize=1000)
+
+                        async def consume(x):
+                            try:
+                                async for item in proxy.syncSplices(x):
+                                    await q.put(item)
+                            finally:
+                                await q.put(None)
+
+                        proxy.schedCoro(consume(offs))
+
+                        done = False
+                        while not done:
+
+                            # get the next item so we maybe block...
+                            item = await q.get()
+                            if item is None:
+                                break
+
+                            items = [item]
+
+                            # check if there are more we can eat
+                            for i in range(q.qsize()):
+
+                                nexi = await q.get()
+                                if nexi is None:
+                                    done = True
+                                    break
+
+                                items.append(nexi)
+
+                            for spliceoffs, item in items:
+                                await self.storNodeEdits(item, {})
```

Comment:
This would be a good spot to integrate a non-lift storNodeEdits() API variant as discussed for migration as well